### ISSUE TITLE
Questions about use, not working for me yet.

### ISSUE THREAD

**Issue by psnijders** [fcrespel/nexus-cas-plugin#1]
Hello Fabien,
I have a question about this plugin. It is very well possible I am doing someting wrong or that I am expecting the wrong things so please bare with me ;-).

1) When I open the star page of Nexus, the CAS plugin tries to log in with user 'anonymous', also when I am already logged in into CAS. The log shows an error 400 (Bad Request, which means: unknown user). Should Nexus not automatically login with the current CAS user ?

2) When I login with a known CAS user (existing in CAS and Nexus) it will not be accepted. The log shows that the CAS authentication goes well, but that Nexus is unable to find permissions. The user, however, exists and has the Nexus administrator role. When I remove the CAS realm from the server settings I can login with this user. Is my user configured wrong ?

jvm 1    | 2013-07-18 18:09:56 DEBUG [nxevthost-1-thread-23]  org.sonatype.nexus.feeds.record.NexusAuthenticationEventInspector - Successfully authenticated user [portal] from IP address 10.55.30.41
jvm 1    | 2013-07-18 18:09:56 DEBUG [qtp17431955-201]  org.sonatype.security.usermanagement.xml.SecurityXmlUserManager - No user role mapping found for user: portal
jvm 1    | 2013-07-18 18:09:56 DEBUG [qtp17431955-201]  org.sonatype.sisu.goodies.eventbus.internal.DefaultEventBus - Event 'org.sonatype.nexus.auth.NexusAuthorizationEvent@1317bfb' fired
jvm 1    | 2013-07-18 18:09:56 DEBUG [nxevthost-1-thread-35]  org.sonatype.nexus.feeds.record.NexusAuthorizationEventInspector - Unable to authorize user [portal] for read(HTTP method "GET") to /nexus/service/local/authentication/login from IP Address 10.55.30.41, user agent:"n/a"
jvm 1    | 2013-07-18 18:09:56 DEBUG [qtp17431955-201]  org.sonatype.nexus.security.filter.authc.NexusSecureHttpAuthenticationFilter - Request processing is rejected because user "portal" lacks permissions.

Thanks for an answer & regards,
Paul.

**Comment by fcrespel** [comment]
Hello,

First, thank you for testing this plugin :-)

1) This plugin does not perform automatic authentication (aka CAS "gateway" mode), as the authentication system used in Nexus doesn't seem flexible enough to easily implement it. For the same reason, the login dialog displayed is still that of the Nexus interface, not the usual CAS page. But I would like to have this feature too, so I may look into it at some point if time permits.

2) This is a use case I haven't tested, so there may be something wrong in the plugin. Personally, I have LDAP configured in addition to CAS with the same backing user directory, so I give permissions to groups and users via the "External role mapping" and "External user role mapping" menus of Nexus. The generated users and roles then appear as coming from LDAP, but their permissions are granted to users authenticated by CAS. It's ugly but it works :-)

In your case, I would guess that the users you have in Nexus are just plain Nexus users, not "External user role mappings", so the permissions you gave them are not granted to CAS users. What is probably missing in the plugin is the ability to create these external mappings for CAS users. I don't know how to do this yet in the interface, but in the mean time here is something you could try :
- When Nexus starts up (with DEBUG log level), look for a line mentioning the internal name of the CAS realm (e.g. org.sonatype.nexus.plugins.cas.CasAuthenticatingRealm_2 - the number may be different for you).
- Edit the "security.xml" file in your "conf" folder.
- In the "userRoleMappings" section, try manually adding a mapping for your test user, using the internal name you found just before as the "source" value.
- (Re)start Nexus, see if it works and let me know. :-)

Fabien.

**Comment by psnijders** [comment]
Well, thank you for making this testing possible in the first place ;-).

I tried your 'security.xml' suggestion but it did not work. I am not sure if I can find the name of the CAS realm though. Grepping my log file for 'CasAuthenticatingRealm' gives this:

```
2013-07-30 12:51:06 DEBUG [jetty-main-thread-1]  org.sonatype.guice.nexus.scanners.NexusTypeVisitor - Found legacy plexus component: org.sonatype.nexus.plugins.cas.CasAuthenticatingRealm
2013-07-30 12:51:06 DEBUG [jetty-main-thread-1]  org.sonatype.guice.nexus.scanners.NexusTypeVisitor - Source: jar:file:/opt/nexus-2.5.0-04/../sonatype-work/nexus/plugin-repository/nexus-cas-plugin-1.0.2-SNAPSHOT/nexus-cas-plugin-1.0.2-SNAPSHOT.jar!/org/sonatype/nexus/plugins/cas/CasAuthenticatingRealm.class
2013-07-30 12:51:06 DEBUG [jetty-main-thread-1]  com.google.inject.internal.BytecodeGen - Loading class org.sonatype.nexus.plugins.cas.CasAuthenticatingRealm FastClass with com.google.inject.internal.BytecodeGen$BridgeClassLoader@224002
2013-07-30 12:51:08 INFO  [jetty-main-thread-1]  org.sonatype.nexus.plugins.cas.CasAuthenticatingRealm - CAS plugin configured for use with server https://test.smart-java.nl/cas
2013-07-30 12:51:08 INFO  [jetty-main-thread-1]  org.sonatype.nexus.plugins.cas.CasAuthenticatingRealm - CAS plugin configured for use with ticket url https://test.smart-java.nl/cas/v1/tickets/
2013-07-30 12:51:08 DEBUG [jetty-main-thread-1]  org.apache.shiro.realm.AuthorizingRealm - CacheManager [org.apache.shiro.cache.ehcache.EhCacheManager@4b35d5] has been configured.  Building authorization cache named [org.sonatype.nexus.plugins.cas.CasAuthenticatingRealm.authorizationCache]
2013-07-30 12:51:08 INFO  [jetty-main-thread-1]  org.apache.shiro.cache.ehcache.EhCacheManager - Cache with name 'org.sonatype.nexus.plugins.cas.CasAuthenticatingRealm.authorizationCache' does not yet exist.  Creating now.
2013-07-30 12:51:08 DEBUG [jetty-main-thread-1]  net.sf.ehcache.store.MemoryStore - Initialized net.sf.ehcache.store.NotifyingMemoryStore for org.sonatype.nexus.plugins.cas.CasAuthenticatingRealm.authorizationCache
2013-07-30 12:51:08 DEBUG [jetty-main-thread-1]  net.sf.ehcache.Cache - Initialised cache: org.sonatype.nexus.plugins.cas.CasAuthenticatingRealm.authorizationCache
2013-07-30 12:51:08 DEBUG [jetty-main-thread-1]  net.sf.ehcache.config.ConfigurationHelper - CacheDecoratorFactory not configured for defaultCache. Skipping for 'org.sonatype.nexus.plugins.cas.CasAuthenticatingRealm.authorizationCache'.
2013-07-30 12:51:08 INFO  [jetty-main-thread-1]  org.apache.shiro.cache.ehcache.EhCacheManager - Added EhCache named [org.sonatype.nexus.plugins.cas.CasAuthenticatingRealm.authorizationCache]
```

So I tried "org.sonatype.nexus.plugins.cas.CasAuthenticatingRealm" as source but that does nothing.

But since my CAS is also configured to use LDAP, theoretically I would be able to do what you did with the external role mappings. This is (only) a matter of configurating LDAP in Nexus and adding the external mappings ? Roles are not read from LDAP (as I understand it) ?

With regard to automatic authentication: I use Nexus proxying through an Apache web server. This makes it possible to secure the Nexus url with a CAS module (mod_auth_cas.so) so you have to log in into CAS to be able to reach Nexus. This ensures that a CAS TGT cookie is present which can then perhaps be read by the plugin to authenticate the user. Just a quick thought ...

**Comment by fcrespel** [comment]
I just committed a change that should allow you to create External mappings for users and roles, without checking if they exist or not (CAS doesn't have any API to do this). This way you shouldn't have to configure LDAP in Nexus to be able to give rights to your users. Would you mind compiling the current head, test it and let me know if it works? :-)

Regarding automatic authentication with mod_auth_cas, it might be possible to make use of it by checking the REMOTE_USER header (which isn't really specific to CAS)... but I don't think this kind of authentication is suitable for Nexus, because client applications (like Maven on the developer's machine, Jenkins on a build server, etc.) need to be able to authenticate programmatically with a simple username and password - they know nothing of CAS. So automatic auth needs to happen only when using the web interface, maybe with some JavaScript code...

**Comment by psnijders** [comment]
Yes, I can definitely confirm that the change works. Thanks !

You are probably right about the auto authentication. To be honest, I do not have much experience with the use of Nexus. When I do I will let you know if I think it is still a good idea ;-).

**Comment by fcrespel** [comment]
Good news then :-)

I just released version 1.1.0 of the plugin with these changes. Thanks for testing it !
